### PR TITLE
Issue 284 Add embed button on embeddable links

### LINKED ISSUE #284: Having to convert an embeddable URL into a hash link for click-to-embed behavior is clumsy
Render a small button/icon beside each embeddable link. Clicking the button should embed the content.

### PR DESCRIPTION
**Pull request by Riott**
Adds a small button to embed on links that have an embeddable link.
image

Resolves #284 

Draft while I get feedback

### CODE REVIEW

File: assets/chat/js/chat.js [hunk 1]
```diff
@@ -1864,7 +1864,10 @@ class Chat {
   cmdEMBED(parts) {
     const { location } = window.top || window.parent || window;
     try {
-      location.hash = this.hashLinkConverter.convert(parts[0]);
+      location.hash = this.hashLinkConverter.convert(
+        parts[0],
+        /* throwErrors */ true
```
Comment (11k) [inline assets/chat/js/chat.js:1869]: Would you mind removing these comments? I think it's a little clumsy and doesn't offer very much utility.

File: assets/chat/js/hashlinkconverter.js [hunk 2]
```diff
@@ -13,8 +13,8 @@ class HashLinkConverter {
     this.rumbleEmbedRegex = /^embed\/([a-z0-9]+)\/?$/;
   }
 
-  convert(urlString) {
-    if (!urlString) {
+  convert(urlString, throwErrors) {
```
Comment (11k) [inline assets/chat/js/hashlinkconverter.js:16]: I'd rather callers who don't care about exceptions simply catch and ignore. I think that's a better approach.

File: assets/chat/js/formatters/UrlFormatter.js [hunk 3]
```diff
@@ -71,14 +72,20 @@ export default class UrlFormatter {
...
         let urlText = encodedUrl;
         if (urlText.length > maxUrlLength) {
           urlText = `${urlText.slice(0, 40)}...${urlText.slice(-40)}`;
         }
         const extra = self.encodeUrl(decodedUrl.substring(m[0].length));
         const href = `${scheme ? '' : 'http://'}${encodedUrl}`;
-        return `<a target="_blank" class="externallink ${extraclass}" href="${href}" rel="nofollow">${urlText}</a>${extra}`;
+        return embedHashLink
+          ? `<a target="_blank" class="externallink ${extraclass}" href="${href}" rel="nofollow">${urlText}</a><button class="embed-button" onclick="location.href='/bigscreen${embedHashLink}'">Embed</button>`
+          : `<a target="_blank" class="externallink ${extraclass}" href="${href}" rel="nofollow">${urlText}</a>${extra}`;
```
Comment (11k) [inline assets/chat/js/formatters/UrlFormatter.js:88]: Would you mind using an anchor over a button? That would make the click handler unnecessary.

**Review by 11k — CHANGES_REQUESTED** [review]
I think a TV icon is fine. You can use an inline SVG.

**Comment by Riott** [comment]
I'll look at addressing these this week if not today at some point :)

File: assets/chat/js/hashlinkconverter.js (hunk 2 above)
Comment (Riott) [reply assets/chat/js/hashlinkconverter.js]: Agreed

File: assets/chat/js/formatters/UrlFormatter.js (hunk 3 above)
Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js]: Yep, no worries

File: assets/chat/js/formatters/UrlFormatter.js [hunk 4]
```diff
@@ -72,14 +73,23 @@ export default class UrlFormatter {
...
         let urlText = normalizedUrl;
         if (urlText.length > maxUrlLength) {
           urlText = `${urlText.slice(0, 40)}...${urlText.slice(-40)}`;
         }
         const extra = self.encodeUrl(decodedUrl.substring(m[0].length));
         const href = `${scheme ? '' : 'http://'}${normalizedUrl}`;
-        return `<a target="_blank" class="externallink ${extraclass}" href="${href}" rel="nofollow">${urlText}</a>${extra}`;
+        return embedHashLink
+          ? `<a target="_blank" class="externallink ${extraclass}" href="${href}" rel="nofollow">${urlText}</a><a class="embed-button" href="/bigscreen${embedHashLink}"><svg xmlns="http://www.w3.org/2000/svg" width="18" height="14.4" viewBox="0 0 640 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc. --><path d="M64 64V352H576V64H64zM0 64C0 28.7 28.7 0 64 0H576c35.3 0 64 28.7 64 64V352c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V64zM128 448H512c17.7 0 32 14.3 32 32s-14.3 32-32 32H128c-17.7 0-32-14.3-32-32s14.3-32 32-32z"  fill="#fff"/></svg></a>`
```
Comment (11k) [inline assets/chat/js/formatters/UrlFormatter.js:91]: No `target` on this anchor means that, when the chat is embedded in an iframe, the stream content is added to that iframe. Merge in master and check out `EmbedUrlFormatter` for how to set this up properly.

**Review by 11k — CHANGES_REQUESTED** [review]

File: assets/chat/js/formatters/UrlFormatter.js (hunk 4 above)
Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js]: I've added in the `const target = chat.isBigscreenEmbed() ? '_top' : '_blank';` line and used that in the anchor link for the embed is there anything additional needed?

File: assets/chat/js/formatters/UrlFormatter.js [hunk 5]
```diff
@@ -90,7 +90,7 @@ export default class UrlFormatter {
...
+          ? `<a target="_blank" class="externallink ${extraclass}" href="${href}" rel="nofollow">${urlText}</a><a target="${target}" class="embed-button" href="https://www.destiny.gg/bigscreen${embedHashLink}"><svg xmlns="http://www.w3.org/2000/svg" width="18" height="14.4" viewBox="0 0 640 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc. --><path d="M64 64V352H576V64H64zM0 64C0 28.7 28.7 0 64 0H576c35.3 0 64 28.7 64 64V352c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V64zM128 448H512c17.7 0 32 14.3 32 32s-14.3 32-32 32H128c-17.7 0-32-14.3-32-32s14.3-32 32-32z"  fill="#fff"/></svg></a>`
```
Comment (11k) [inline assets/chat/js/formatters/UrlFormatter.js:93]: The full URL should use `chat.config.dggOrigin` and `chat.bigscreenPath` for the origin, just like in `EmbedUrlFormatter`. Also, would you mind getting rid of the license comment for the SVG?

**Review by 11k — CHANGES_REQUESTED** [review]

File: assets/chat/js/formatters/UrlFormatter.js (hunk 5 above)
Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js]: Sorry about that, resolved, removed the license too!.

**Review by 11k — APPROVED** [review]

**Pull request merged**